trio/testing/_memory_streams.py

```diff
@@ -67,17 +67,19 @@ def _get_impl(self, max_bytes):
             return bytearray()
 
     def get_nowait(self, max_bytes=None):
-        with self._fetch_lock.sync:
+        with self._fetch_lock:
             self._check_max_bytes(max_bytes)
             if not self._closed and not self._data:
                 raise _core.WouldBlock
             return self._get_impl(max_bytes)
 
     async def get(self, max_bytes=None):
-        async with self._fetch_lock:
+        with self._fetch_lock:
             self._check_max_bytes(max_bytes)
             if not self._closed and not self._data:
                 await self._lot.park()
+            else:
+                await _core.checkpoint()
             return self._get_impl(max_bytes)
```

Comment:
I think this has the same problem as `UnboundedQueue` – how do we know that after we wake up, `self._data` won't have disappeared?

Reply:
Well, anyone else consuming `self._data` would have to acquire the `_fetch_lock` ConflictDetector, which can't happen while we're holding it. I agree that it looks suspicious, though, so have cleaned it up.

Reply:
Oh yeah... in fact if this was a problem then the `park` version would have the same problem, because someone could steal the buffer in between the call to `unpark` and when the task actually woke up. And github helpfully hid the `with self._fetch_lock` line from me, which is the part that *actually* gives you critical information that makes it safe :-)

I'm going to revert back to what you had before your last commit...